tests/python/pants_test/rules/test_filedeps.py

```diff
@@ -0,0 +1,282 @@
+# coding=utf-8
+# Copyright 2018 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+import os
+
+from future.utils import text_type
+from pex.orderedset import OrderedSet
+
+from pants.build_graph.address import Address, BuildFileAddress
+from pants.engine.addressable import BuildFileAddresses
+from pants.engine.fs import PathGlobsAndRoot, PathGlobs
+from pants.engine.legacy.graph import HydratedTarget, TransitiveHydratedTargets
+from pants.engine.legacy.structs import TargetAdaptor
+from pants.rules.core.filedeps import file_deps
+from pants.util.contextutil import temporary_dir
+from pants.util.dirutil import safe_file_dump
+from pants.util.meta import AbstractClass
+from pants.util.objects import datatype
+from pants_test.engine.scheduler_test_base import SchedulerTestBase
+from pants_test.engine.util import MockConsole, run_rule
+from pants_test.test_base import TestBase
+
+
+class FileDepsTest(TestBase, SchedulerTestBase, AbstractClass):
+
+  def filedeps_rule_test(self, target_addresses, transitive_targets, expected_console_output):
+    console = MockConsole()
+
+    run_rule(file_deps, console, BuildFileAddresses(target_addresses), {
+      (TransitiveHydratedTargets, BuildFileAddresses): lambda _: transitive_targets,
+    })
+
+    self.assertEquals(console.stdout.getvalue(), expected_console_output)
+
+  def make_build_target_address(self, spec):
+    address = Address.parse(spec)
+    
+    return BuildFileAddress(
+      build_file=None,
+      target_name=address.target_name,
+      rel_path='{}/BUILD'.format(address.spec_path),
+    )
+
+  def make_build_target_address_with_ext(self, spec):
+    address = Address.parse(spec)
+
+    return BuildFileAddress(
+      build_file=None,
+      target_name=address.target_name,
+      rel_path='{}/BUILD.ext'.format(address.spec_path),
+    )
+
+  def make_snapshot(self, files):
+    with temporary_dir() as temp_dir:
+      for file_name, content in files.items():
+        safe_file_dump(os.path.join(temp_dir, file_name), content)
+      return self.scheduler.capture_snapshots((
+        PathGlobsAndRoot(PathGlobs(('**',)), text_type(temp_dir)),
+      ))[0]
+    
+  def mock_hydrated_target(self, target_address, source_files_in_dict, dependencies):
```

Comment:
@stuhood Would appreciate any ideas you have for mocking this whole nested structure of a `TargetAdaptor` better than this...

Reply:
I'd suggest using python's mock library, and exposing the bare minimum of fields you need in order to pretend to be a `HydratedTarget`. This looks like a good example:
https://github.com/pantsbuild/pants/blob/master/tests/python/pants_test/backend/jvm/tasks/test_jar_publish.py#L104-L113

Reply:
But also, given that `HydratedTarget` and `TransitiveHydratedTarget` are not "forever" (see #4535), biasing _slightly_ more toward integration tests might be good.